### ISSUE TITLE
AllGather incorrect output / hangs for certain test cases in dram

### ISSUE THREAD

**Issue by tt-aho** [tenstorrent-metal/tt-metal#5396]
Seeing weird behaviour of all gather getting mismatches/hangs.

Note that this is the older implementation of all gather. The newer one where sender always pulls from memory each hop has not been tested with these cases yet.

The most common case are RM tensors in DRAM, where going above 13 rows of input I start seeing these issues.
There is also one hang case if Tile layout in DRAM when num_links=2.

It seems like some of these hang/mismatch cases can go away when introducing a delay after the sender core reads from dram.

Debug branches/tests:
`aho/all-gather-debug` and `aho/all-gather-debug2`
The first one has everything stripped out of all gather except reading data into l1 and dprinting a couple values. I'm seeing that the second value dprinted is garbage, but it’s correct if I write to erisc_info and readback or remove the first dprint or add some delays after the async_reads or even after the dprint? Delay fixing after dprint is very strange.

Second one is full all gather test with a dprint/added delay after read barrier.
If you increase input rows above 12 you start seeing incorrect output/hangs. If I dprint after the barrier output for 13 rows is correct, but 33 still hangs. If I add a delay after then 13 has incorrect output, but 33 now passes
For both of these the test is `tests/tt_eager/python_api_testing/unit_testing/test_all_gather.py `and kernel is `tt_eager/tt_dnn/op_library/all_gather/kernels/dataflow/interleaved_eth_ring_gather_send_stick_layout.cpp`

**Comment by tt-aho** [comment]
For falcon we are always pulling from L1 so there haven't been any issues so far.

**Comment by SeanNijjar** [comment]
I don't know if any of this applies anymore with the most recent implementation. 

What I can say is I did sometimes see a hang where one of the worker cores never receives its "go" signal and so it sits idle. Moving the workers around gets around the issue. It wasn't root caused.

I also noticed an occasional non-deterministic PCC issue. It looks like it might be an issue with the API somewhere? Still needs to be root caused but some interesting details I noticed is it's only the last part of the tensor that ever mismatches, never the beginning or middle. Additionally, I only ever see a mismatch for Tensor 0 (the one on chip 0). I was seeing this with bidirectional, which means that it's the last tensor data that should be committed to DRAM that's mismatching. It's as if the write barriers aren't actually waiting for the payload to land in DRAM.
